src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -126,10 +113,35 @@ impl Qualifs<'a, 'mir, 'tcx> {
 
         let return_loc = item.body.terminator_loc(return_block);
 
+        let custom_eq = match item.const_kind() {
+            // We don't care whether a `const fn` returns a value that is not structurally
+            // matchable. Functions calls are opaque and always use type-based qualification, so
+            // this value should never be used.
+            ConstKind::ConstFn => true,
+
+            // If we know that all values of the return type are structurally matchable, there's no
+            // need to run dataflow.
+            ConstKind::Const | ConstKind::Static | ConstKind::StaticMut
+                if !CustomEq::in_any_value_of_ty(item, item.body.return_ty()) =>
+            {
+                false
+            }
```

Comment:
I wonder if every qualif could benefit from an approach like this? Oh but within a `const fn` we use the other qualifs for const-checking, nevermind.

Reply:
The `CustomEq` qualif is also a lot slower than the others because the functions in `structural_match.rs` aren't cached and have to create a new inference context for each call. If these changes don't speed things up enough, caching `search_for_structural_match_violation` and `type_marked_structural` is the next step.